Scala 2.12 upgrade preview with mysql tests fixed

pankajroark opened the pull request:
Builds on top of #342 fixing the mysql tests that the scalacheck upgrade breaks.

Commit:
Upgrades to Scala 2.12.

1. Incremented all required versions to value that included scala 2.12 support. See diff for details.
2. Modified RedisStore to use Buf instead of ChannelBuffer due to changes in underlying client. To ease this
   transition a method to convert Store[Buf, Buf] to Store[ChannelBuffer, ChannelBuffer]
   is provided on RedisStore.
3. Removed support for Scala 2.10. Two libraries lacked support for scala 2.10 and 2.12 on the same version: twiiter-util and finagle.
   Supporting 2.10 finagle and 2.12 finagle was not possible due to interface changes.
4. Various packages changed and were fixed.
5. Added target to travis.yml and build.sbt to include scala 2.12.
6. Set source and build target to java 8 to support scala 2.12.

pankajroark committed:
Fix mysql tests that break with scalacheck 1.13.4

pankajroark commented:
All tests have passed afaict, looks like travis is having issues.

pankajroark committed:
Update finagle and twitter/util versions.

ttim commented on .travis.yml:
```diff
@@ -2,18 +2,24 @@ language: scala
-      script: umask 0022 && ./sbt ++$TRAVIS_SCALA_VERSION clean test mimaReportBinaryIssues
-
     - scala: 2.11.8
+      jdk: oraclejdk8
```
Should we do this for 2.11?

pankajroark replied:
No, I guess in the original PR Ben may have added it for consistency. Let me remove it.

ttim commented on build.sbt:
```diff
@@ -65,9 +65,9 @@ val ignoredABIProblems = {
-  crossScalaVersions := Seq("2.10.6", "2.11.7"),
-  javacOptions ++= Seq("-source", "1.6", "-target", "1.6"),
-  javacOptions in doc := Seq("-source", "1.6"),
+  crossScalaVersions := Seq("2.11.7", "2.12.1"),
```
If we add 2.12 here then should we also enable travis for 2.12?

johnynek replied:
can we use `2.11.11` and `2.12.2`?

ttim commented on storehaus-algebra/src/test/scala/com/twitter/storehaus/algebra/query/CalendarTimeQueryRangeProperties.scala:
```diff
@@ -19,7 +19,8 @@ package com.twitter.storehaus.algebra.query
 import org.scalacheck.Prop._
 import org.scalacheck.Properties
-import com.twitter.scalding._
+import com.twitter.scalding.{RichDate, DateRange, Duration, AbsoluteDuration, Years,
```
Do we need it in this review?

johnynek replied:
+1 to minimize. I prefer to keep the diffs as minimal as possible (and personally never mix style and logic changes if possible).

pankajroark replied:
Removed.

ttim reviewed (approved):
Looks good overall, the only questionable thing for me - should we enable 2.12 as part of this change?

johnynek commented on .travis.yml:
```diff
@@ -2,18 +2,24 @@ language: scala
       script: umask 0022 && ./sbt ++$TRAVIS_SCALA_VERSION clean coverage test coverageReport mimaReportBinaryIssues
       after_success:
         - bash <(curl -s https://codecov.io/bash)
+#    - scala: 2.12.1
```
what's the story here?

pankajroark replied:
Ahh... this got carried over from Ben's original PR. Let me uncomment this.

johnynek commented on storehaus-memcache/src/main/scala/com/twitter/storehaus/memcache/MemcacheStore.scala:
```diff
@@ -70,7 +70,6 @@ object MemcacheStore {
       .tcpConnectTimeout(timeout)
       .requestTimeout(timeout)
       .connectTimeout(timeout)
-      .readerIdleTimeout(timeout)
```
this seems like a maybe logical change. Are we sure there is nothing to replace it?

pankajroark replied:
Applied the alternative way of doing this.

johnynek commented on storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala:
```diff
@@ -38,34 +38,34 @@ object RedisSetStore {
       case (k, Some(v)) =>
-        client.del(Seq(k)) // put, not merge, semantics
-        ttl.map(exp => client.expire(k, exp.inSeconds))
+        client.dels(Seq(k)) // put, not merge, semantics
```
we don't need to wrap with a `Seq`. Could we just call `client.del(k)` maybe? Also, this seems like a a race to me. the del could happen after the set. We should use flatMap to sequence:
```scala
for {
  _ <- client.del(k)
  _ <- ttl.fold(Future.Unit) { exp => client.expire(k, exp.inSeconds) }
  _ <- set(k, v.toList)
} yield ()
```

pankajroark replied:
del is deprecated unfortunately. Taken care of the race.

johnynek commented on storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala:
```diff
@@ -38,34 +38,34 @@ object RedisSetStore {
         set(k, v.toList)
       case (k, None) =>
-        client.del(Seq(k)).unit
+        client.dels(Seq(k)).unit
```
is there no`client.del(k)` we can call?

pankajroark replied:
del is deprecated:
  @deprecated("remove netty3 types from public API", "2016-03-24")
  def del(keys: Seq[ChannelBuffer]): Future[JLong] =

johnynek commented on storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala:
```diff
@@ -32,35 +32,35 @@ object RedisSortedSetStore {
-  override def get(k: ChannelBuffer): Future[Option[Seq[(ChannelBuffer, Double)]]] =
-    client.zRange(k, 0, -1, true).map(
+  override def get(k: Buf): Future[Option[Seq[(Buf, Double)]]] =
+    client.zRange(k, 0.toLong, -1.toLong, true).map(
```
`0L` and `-1L` is the better way to do this.

johnynek commented on storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala:
```diff
@@ -32,35 +32,35 @@ object RedisSortedSetStore {
     kv match {
       case (set, Some(scorings)) =>
-        client.del(Seq(set)).flatMap { _ =>
+        client.dels(Seq(set)).flatMap { _ =>
```
is there no `client.del(set)`?

pankajroark replied:
There is but it's marked deprecated.

johnynek commented on storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala:
```diff
@@ -32,35 +32,35 @@ object RedisSortedSetStore {
         }
       case (set, None) =>
-        client.del(Seq(set)).unit
+        client.dels(Seq(set)).unit
```
same comment about `client.del(set)`.

pankajroark replied:
del is deprecated.

johnynek commented on storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisStore.scala:
```diff
@@ -34,24 +40,33 @@ object RedisStore {
+  }
+}
+
+object ChannelBuffer2BufInjection extends Injection[ChannelBuffer, Buf] {
```
this should be a bijection, not an Injection (there is an implicit from Bijection to Injection, by the way so you can pass this as the injection above).

pankajroark committed:
Take care of review comments. Going back to finagle 6.41.
finagle-http-compat has been removed in more recent versions, taking
care of that may need non-trivial changes. Will
upgrade finagle version in a follow up change.

pankajroark committed:
Use jdk8 for scala2.11 travis.

pankajroark committed:
Mongodb tests were going OOM due to creation of client numerous times, again due to property declaration behavior change in scalacheck. Fix that.

pankajroark committed:
Remove mima checks for 2.12 for this change as this is the first release for 2.12

pankajroark committed:
Add excludes for mima failures.

oscar-stripe commented:
+1

After we publish a new version, we should bump the minor version, and remove all the exclusions for mima.

Thanks for doing this.
